beacon-chain/blockchain/forkchoice_update_execution.go

```diff
@@ -52,6 +52,44 @@ type fcuConfig struct {
 	attributes    payloadattribute.Attributer
 }
 
+// sendFCU handles the logic to notify the engine of a forckhoice update
+// for the first time when processing an incoming block during regular sync. It
+// always updates the shuffling caches and handles epoch transitions when the
+// incoming block is late, preparing payload attributes in this case while it
+// only sends a message with empty attributes for early blocks.
+func (s *Service) sendFCU(cfg *postBlockProcessConfig, fcuArgs *fcuConfig) error {
+	if !s.isNewHead(cfg.headRoot) {
+		return nil
+	}
+	if fcuArgs.attributes != nil && s.shouldOverrideFCU(cfg.headRoot, s.CurrentSlot()+1) {
+		return nil
+	}
+	return s.forkchoiceUpdateWithExecution(cfg.ctx, fcuArgs)
+}
+
+// sendFCUWithAttributes computes the payload attributes and sends an FCU message
+// to the engine if needed
+func (s *Service) sendFCUWithAttributes(cfg *postBlockProcessConfig, fcuArgs *fcuConfig) {
+	if !s.isNewHead(cfg.headRoot) {
+		return
+	}
+	slotCtx, cancel := context.WithTimeout(context.Background(), slotDeadline)
+	defer cancel()
+	cfg.ctx = slotCtx
```

Comment:
Here we send the modified context to all receivers. Notice however that we are modifying the parent's context of `postBlockProcess`. This is currently fine cause that context is not used anywhere else before calling this function and `postBlockProcess` ended right after calling this function. However, in the future it may be better to collect handlers to be called at the end of `ReceiveBlock` instead.